Consolidate and clean up the scripting section

mattgarrish opened the pull request:
This PR makes the following changes:

- creates a new section called "script inclusion" that includes the text about adding scripts to content documents (removes the scripting context definitions)
- adds an explanation about origins and the effect on scripting (no data sharing) to the inclusion section
- creates a new "scripting contexts" section with a short intro followed by subsections that cover each of the contexts in detail (basically consolidates prose that was spread across sections before)
- creates a new appendix for detailed examples and moves the context example to it (adding together with the example of a packaged epub and wall clock values so we avoid a proliferation of top-level appendixes)

Mostly shuffling with some rewriting, in other words, but hopefully a little easier to read now.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/epub-specs/pull/1519.html" title="Last updated on Feb 24, 2021, 2:23 PM UTC (bd562a3)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/epub-specs/1519/b9642ea...bd562a3.html" title="Last updated on Feb 24, 2021, 2:23 PM UTC (bd562a3)">Diff</a>

mattgarrish committed:
consolidate and clean up the scripting section

mattgarrish committed:
move example to appendix and consolidate with other detailed examples

iherman commented on epub33/core/index.html:
```diff
@@ -4362,123 +4320,118 @@ <h4>Scripting Contexts</h4>
 							a future update adds the concept.</p>
```
This is purely aesthetic: I would merge the two notes in one, and let the sentence on the data block follow the initial paragraph. The green boxes for notes break the flow...

mattgarrish replied:
Merging doesn't really make sense in this case as the notes refer to two separate statements. It would jumble the discussion.

The first note could probably be regular spec prose, though.

iherman commented on epub33/core/index.html:
```diff
@@ -4362,123 +4320,118 @@ <h4>Scripting Contexts</h4>
-							Conformance</a> [[EPUB-RS-33]] for some specific requirements that have to be adhered to
-						(not all Reading Systems will provide the same scripting functionality).</p>
+					<p>Authors should note that Reading Systems are expected to behave as though a unique <a
+							href="https://url.spec.whatwg.org/#origin">origin</a> [[URL]] has been assigned to each EPUB
```
I know this is in line with the change on the origin, but I would prefer to see a stronger term than "expected", ie, "required".

iherman commented on epub33/core/index.html:
```diff
@@ -4362,123 +4320,118 @@ <h4>Scripting Contexts</h4>
+							instructions for modifying the DOM of the EPUB Content Document that embeds it (i.e., the
+							one that contains the <code>iframe</code> element). It also MUST NOT contain instructions
+							for manipulating the size of its containing rectangle.</p>
 
```
The problem with these statements is that I do not see how this can be tested. I presume epubcheck would be unable to do that, it would require to understand the behavior of the script.

I am completely out of my comfort zone here, maybe what I say is very different. Looking at the [iframe `sandbox` attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe) isn't it possible to replace this (or make it more precise) by applying constraints on the attributes the `<iframe>` element should or should not use? _That_ can be tested by epubcheck.

I am just wondering; maybe @rdeltour, @danielweck, or @bduga can comment on this...

mattgarrish replied:
Can you open a separate issue for this as it would change existing normative requirements?

I don't want to open a new discussion in a PR.

iherman replied:
You are right. I will.

pinging @dauwhe and @wareid because we just had some discussions about this...

iherman replied:
done in #1523

iherman commented on epub33/core/index.html:
```diff
@@ -4362,123 +4320,118 @@ <h4>Scripting Contexts</h4>
+								spine-level scripting in Reading Systems</a> is only recommended in fixed layout
+							documents and reflowable documents when they are set to scroll [[EPUB-RS-33]]. Furthermore,
+							Reading System support in all other contexts is optional.</p>
+
```
I wonder whether a more direct reference to [§4.4.1](https://w3c.github.io/epub-specs/epub33/rs/#sec-scripted-content-rs-reqs) in the RS document, that is much more precise about by the properties that must be used. Ie, statements like:

> reflowable EPUB Content Documents that use the "scrolled-doc" or "scrolled-continuous" [EPUB-33] presentation modes defined by the rendition:flow property.

are much more concrete and enforceable, ie, the author should use those constraints.

That being said, I am a bit bothered by this restriction (I know, that may be an inherited feature that we cannot change). For example, a script that puts a line number to each line in a `<pre>` element does not fall into this category if the content is displayed in paginated mode, although that may make a lot of sense. I understand that the danger is to interfere with the pagination algorithm of the RS, but I still wonder whether we can be, somehow, more liberal...

mattgarrish replied:
> I understand that the danger is to interfere with the pagination algorithm of the RS, but I still wonder whether we can be, somehow, more liberal...

Again, this is probably one for a new issue.

iherman replied:
True for the last remark, and maybe we should simply ignore this for now...

mattgarrish committed:
makes the following changes per feedback:
remove note formatting;
s/expected/required/;
add links to script support bullets for spine scripting
